siliconcompiler/report/report.py

```diff
@@ -294,12 +285,57 @@ def get_files(chip, step, index):
         step (string) : Step of node.
         index (string) : Index of node.
     """
-    # could combine filters, but slightly more efficient to separate them
+    # could combine filters, but slighlty more efficient to seperate them
     # Is remaking the list with sets instead of list worth it?
     logs_and_reports = []
     all_paths = os.walk(chip._getworkdir(step=step, index=index))
     for path_name, folders, files in all_paths:
-        logs_and_reports.append((path_name,
-                                 set(folders),
-                                 set(files)))
+        logs_and_reports.append((path_name, set(folders), set(files)))
     return logs_and_reports
+
+
+def get_chart_data(chips, metric, steps_and_indicies):
+    """
+    Returns returns a a tuple where the first element is a 2d dictionary of
+    data points, following the forms {step+index: {chip_name: value}} where
+    each dictionary can have many keys. The second element is a string
+    (or None) that represents the unit.
+
+    Args:
+        chips (list) : A list of dictionaries with the form
+            {'chip_object': chip, 'chip_name': name}.
+        metric (string) : The metric that the user is searching.
+        steps_and_indicies (list) : A list of dictionaries with the form
+            {'step': step, 'index': index}.
+    """
+    chips_included = set()
+    metric_datapoints = {}
+    metric_units = set()
+    for step_and_index in steps_and_indicies:
+        step = step_and_index['step']
+        index = step_and_index['index']
+        metric_datapoints[(step, index)] = {}
+        for chip_and_chip_name in chips:
+            chip = chip_and_chip_name['chip_object']
+            chip_name = chip_and_chip_name['chip_name']
+            metric_unit = None
+            value = chip.get('metric', metric, step=step, index=index)
+            if value is not None:
+                if chip.schema._has_field('metric', metric, 'unit'):
+                    metric_unit = chip.get('metric', metric, field='unit')
+                metric_type = chip.get('metric', metric, field='type')
+                value = utils._format_value(metric, value, metric_unit, metric_type)
+                if value.isnumeric():
+                    value = int(value)
+                else:
+                    try:
+                        value = float(value)
+                    except (TypeError, ValueError):
+                        pass
+                if metric_unit:
+                    metric_units.add(metric_unit)
+                chips_included.add(chip_name)
+                metric_datapoints[(step, index)][chip_name] = value
```

Comment:
I could, but that would lead to a lot of extra computation with how the `utils._collect_data`  works at the moment. If there are 20 metrics and 10 nodes on 10 chips for a query of 1 metric and 1 node and all 10 chips, the bulk of the computation runs 10 times with how this function is constructed. Your suggestion would cause this to run 2000 times. We could speed it up by getting all the metrics tables at the start and reusing those. I'm not sure which option is better though.

Reply:
cant you call it 10 times for each chip and just index into the matrix that is returned?

Reply:
I think this is probably something we should talk about.